File: golang/infrastructure/httputils/productsFilters.go
package httputils

import (
	"strconv"

	"github.com/Akiles94/mytheresa-test/domain/dto"
	"github.com/Akiles94/mytheresa-test/domain/models"
	"github.com/gin-gonic/gin"
	log "github.com/sirupsen/logrus"
)

func GetProductsFiltered(data []models.Product, filters dto.QueryParams) []models.Product {
	response := []models.Product{}
	for _, item := range data {
		catFlag := false
		priceFlag := false
		if filters.Category != nil {
			if item.Category == *filters.Category {
				catFlag = true
			}
		} else {
			catFlag = true
		}
		if filters.PriceLessThan != nil {
			if item.Price < *filters.PriceLessThan {
				priceFlag = true
			}
		} else {
			priceFlag = true
		}
		if catFlag && priceFlag {
			response = append(response, item)
		}
	}
	return response
}

func GetProductsParams(c *gin.Context) (*dto.QueryParams, error) {
	var result dto.QueryParams
	var category *string
	var priceLessThan *int
	tmpCategory := c.DefaultQuery("category", "")
	if tmpCategory != "" {
		category = &tmpCategory
	}
	tmpPricesLessThan, err := strconv.Atoi(c.DefaultQuery("priceLessThan", "0"))
	if err != nil {
		log.Error("Error converting from string to int for offset")
		return nil, err
	}
	if tmpPricesLessThan != 0 {
		priceLessThan = &tmpPricesLessThan
	}
	result = dto.QueryParams{
		Category:      category,
		PriceLessThan: priceLessThan,
	}

	return &result, nil
}

File: symfony/src/Infrastructure/HttpUtils/ProductsFilters.php
<?php

namespace App\Infrastructure\HttpUtils;

use App\Domain\Dto\QueryParams;
use Symfony\Component\HttpFoundation\Request;

class ProductsFilters {

    public static function GetProductsParams(Request $request):QueryParams{
        $response = null;
        $category = $request->get('category');
        $priceLessThan = $request->get('priceLessThan');        
        $limit = $request->get('limit',20);        
        $offset = $request->get('offset',0);        
        $response = new QueryParams(
            category: $category,
            priceLessThan: $priceLessThan,
            limit: $limit,
            offset: $offset,
        );
        return $response;
    }

    public static function GetProductsFiltered($products, QueryParams $params){
        $response = [];
        $count = 0;
        foreach($products as $product){            
            $count++;
            $catFlag = false;
            $priceFlag = false;
            if(!is_null($params->category)){
                if($params->category == $product->category){
                    $catFlag = true;
                }                
            }else{
                $catFlag = true;
            }
            if(!is_null($params->priceLessThan)){
                if($product->price < $params->priceLessThan){
                    $priceFlag = true;
                }                
            }else{
                $priceFlag = true;
            }
            if($catFlag === true && $priceFlag === true && $count <= $params->limit && $count > $params->offset){
                array_push($response, $product);
            }
        }
        return $response;
    }    
}